deployment/actionManger.go

```diff
@@ -274,12 +282,113 @@ func (mgr *ActionManager) ensureSomeWorkerIsRunning(compID worker.ComponentID) e
 
 	// Update the hash we're storing if we had HEAD
 	if compID.Hash == headHashSentinel {
-		mgr.pathHashes[compPath] = deployedHash
+		mgr.pathHashes[compPath].hash = deployedHash
 	}
 
 	return nil
 }
 
+func (mgr *ActionManager) ensureNWorkerIsRunning(compID worker.ComponentID) error {
+	compPath := worker.ComponentPath{
+		User: compID.User,
+		Repo: compID.Repo,
+	}
+
+	//Get instance states
+	compMap := getCurrentInstanceState(mgr.workers)
+	if _, ok := compMap[compID]; ok {
+		//Check if should scale up
+		if compMap[compID].numInstances < mgr.pathHashes[compPath].numInstances {
+			//Find worker where this comp isn't deployed
+			workerToDeployTo, err := mgr.findWorkerToDeployTo(compID)
+			if err != nil {
+				log.Error.Println("Worker deployed on all nodes")
+				return nil
+			}
+			//Deploy to worker
+			deployedHash, err := mgr.activator.Activate(compID, workerToDeployTo)
+			if err != nil {
+				return err
+			}
+			// Update the hash we're storing if we had HEAD
+			if compID.Hash == headHashSentinel {
+				mgr.pathHashes[compPath].hash = deployedHash
+			}
+			return nil
+		}
+		//Check if should scale down
+		if compMap[compID].numInstances > mgr.pathHashes[compPath].numInstances {
+			//Find worker where this comp is deployed
+			workerToDeactivateOn, err := mgr.findWorkerToDeactivateIOn(compID)
+			if err != nil {
+				log.Error.Println("Comp not running on any workers")
+				return nil
+			}
+			//Deactivate
+			err = mgr.activator.Deactivate(compID, workerToDeactivateOn)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+	} else {
+		//Deploy a single instance
+		workerToDeployTo := mgr.workers[rand.Intn(len(mgr.workers))]
+		deployedHash, err := mgr.activator.Activate(compID, workerToDeployTo)
+		if err != nil {
+			return err
+		}
+		// Update the hash we're storing if we had HEAD
+		if compID.Hash == headHashSentinel {
+			mgr.pathHashes[compPath].hash = deployedHash
+		}
+		return nil
+	}
+	return errors.New("Something weird happened.")
+}
+
+func (mgr *ActionManager) findWorkerToDeployTo(compID worker.ComponentID) (*worker.V9Worker, error) {
```

Comment:
Doesn't this function already exist in this file?

Reply:
This function ensures that the worker it returns doesn't have the component running on it.